### ISSUE TITLE
Feature request: automated install when the required config values are present

### ISSUE THREAD

**Issue by dessant** [piwik/piwik#10257]
If `config.ini.php` is present, the installation sequence is skipped entirely.

When the required fields are present in the config file (database host, name, password etc), the installation could happen in a headless way, otherwise the manual istallation page would be loaded on the first visit, and the entered data would be stored in config *without* overwiting the existing contents of the config file.

This would make it easy to provide a custom config during provisioning and fully set up Piwik, without the need for hacks like these: https://github.com/MarvAmBass/docker-piwik/blob/master/startup-piwik.sh.

I am aware of headless installation being available in the Pro version, please note that this issue was opened for the free version of Piwik.

**Comment by tsteur** [comment]
The installation is skipped indeed if all needed fields are there in the config and if database tables exist. I'm not quite sure what the exact problem is? Did you want to install Piwik via command line by providing parameters like "--database-host"?

**Comment by dessant** [comment]
Yeah, the main issue is that headless installation is not supported and one has to use the ui to provide the db connection and bootstrap the database. While this is ok for one-off installations, it becomes problematic when you're doing automated provisioning.

The config way was one idea, ideally the main outcome would be to have a way to install Piwik up to a state where the admin user is created, without requiring manual steps.

**Comment by tsteur** [comment]
got it. I thought I created a similar issue a few years ago when I built a vagrant box for Piwik but cannot find it anymore. Possibly I didn't actually create the issue. It would be a useful feature to have nowadays. 

Basically we would only need to create the config file from the input and validate the entered data. For validation and creation of config file everything is already here so it shouldn't be too complex

**Comment by mattab** [comment]
Hi @dessant 

we may in the future release such a tool as open source. If your company can support us in this area, please get in touch via https://piwik.org/contact/

**Comment by dessant** [comment]
Thanks for considering a possible release, I'll definitely do it when I will have a company or I will work for one. :yum:

**Comment by Toub** [comment]
Let's say the automation can be run with a script named install-piwik.sh, we need 2 different ways to provide the configuration:
- command lines args (e.g. install-piwik.sh --dbuser=myuser --dbpassword=pass ...) => drawback: the tool appears in the command history
- install.ini file (running install-piwik.sh will then remove install.ini, so it can be more secure)

**Comment by tsteur** [comment]
If someone is keen to work on this I'm happy to assist. Be good to have this

**Comment by nebev** [comment]
I had a stab at this yesterday. It's not particularly brilliant, but gets the job done for me.
I'm running docker containers, so it's not really an option for me to use the CURL method.
For anyone interested: https://github.com/nebev/piwik-cli-setup

**Comment by gaumondp** [comment]
Thanks @nebev , I'm no programmer but I can think of some more features or maybe it should be a different script? 

1. Option to upgrade instead of installing
2. Option to fetch the source from http://builds.piwik.org/latest.tar.gz (or any version) or a local directory.

I'm also wondering if such script should be maintain by Piwik CoreTeam once interest grow?

All in all, thanks anyway for sharing the code. :)

**Comment by lorengordon** [comment]
I'm also very interested in an automated mechanism to complete the initial configuration and creation/update of the db tables. 👍

**Comment by mattab** [comment]
Please note we have this automation available but it's one of the rare things we haven't published as Open source. If you want to access this please contact InnoCraft here: https://www.innocraft.com/#contact

**Comment by pc-magas** [comment]
As for me I wanted to have a cli installer in order to make a custom docker image.

**Comment by chriscroome** [comment]
I have been looking at how to script Piwik installs and came across the [Cauldron install script](https://git.cloudron.io/cloudron/piwik-app/blob/master/start.sh) which adopts another approach to this problem from the examples above -- using a minimal MySQL dump to populate the database and then using the php console script to setup the config file.

**Comment by nebulade** [comment]
Hi, I'm the author of that Cloudron package and I think it is a quite bad hack using the sql dump. Since I lack deeper knowledge of the piwik install process, this was just an option to get around the setup in our case, but I would not recommend that in the long run at all. The alternative solution I did first was to replay the requests to the setup webinterface via script, but that was even uglier.

**Comment by nebev** [comment]
I agree. SQL dump is a bad idea as schemas change. Best to use API calls and existing functions where possible. I'm successfully running my CLI setup script in a docker image for brand new installs of Piwik 3.0.2.

**Comment by tomaluca95** [comment]
I'm interested if anyone has a good solution to automatically install (and maybe upgrade) to the last stable version.
I'm trying to create an ansible playbook to setup the full stack and piwik is part of the environment.

**Comment by jamietanna** [comment]
I've been doing this via Chef, but it uses SQL dumps so as mentioned, isn't great long term. However, it means I can always install the same Piwik version correctly

**Comment by minac** [comment]
Any news on this one? I have everything setup using docker and docker-composer including a customized config.ini.php file. It all works when I do a first manual run that creates the user and database tables, but if config.ini.php is present it fails (as was said above). Could we see matomo create/bootstrap the database when it's not there but the credentials are there?

Thanks!

**Comment by mikkeschiren** [comment]
@minac  I have been working on this the recent days, repo is over here: https://github.com/nodeone/extratools
The plugin adds some new console commands - adn most important matomo:install .

The work is still work in progress - but I hope I have the most parts ready this week and will after that ask to have the plugin published at https://plugins.matomo.org/ if the Matomo people approve.

**Comment by mattab** [comment]
Hi @mikkeschiren - it's great to see your project, giving everyone the ability to install Matomo in one command line... on our end in Matomo, we'd need to have the `install` command in core (and maybe can work together to get it merged into core later), and in the meantime the community will hopefully start using your project :+1:

**Comment by mikkeschiren** [comment]
I will gladly help to get this into core later @mattab. We (Digitalist Sweden) - mainly work with fully automated workflows and Matomo has become a very important part of our business the latest year, so we really want to contribute back.

**Comment by mikkeschiren** [comment]
Still some minor details I need to solve before an official release, but should be just days from now,

**Comment by mikkeschiren** [comment]
First alpha is out now: https://plugins.matomo.org/ExtraTools

**Comment by colans** [comment]
> I'm trying to create an ansible playbook to setup the full stack and piwik is part of the environment.  I'm interested if anyone has a good solution to automatically install (and maybe upgrade) to the last stable version.

In researching the Ansible roles for Matomo that I've found, most of them are downloading the `zip` file, which seems sub-optimal.

I believe the following plan should do the trick (in an Ansible role):
1. Install the Debian package [like this example does](https://mackuba.eu/2019/02/02/setting-up-piwik/) (in its "Installing Piwik" section).
1. Keep the VM up-to-date with [Automatic Updates](https://help.ubuntu.com/lts/serverguide/automatic-updates.html).  This will ensure that the Debian package is kept up-to-date.  (You could do this in your default-VM role.)
1. Run through the manual Web configuration once, which generates `config.ini.php`.  Save that file as a template, which gets injected into the installation, but replace hard-coded values with variables.  Some of the roles I've seen are doing this, but others are injecting a database dump, which I feel is overkill.

Correct me if I'm wrong, but I believe the above solution solves the issue without the need for a console option.  I'm assuming this will also work with Chef, Puppet, Salt, etc.

**Comment by minac** [comment]
Hey @colans, 

That approach seems sound. The only downside is the database upgrades. Matomo has 2 types of upgrades that usually go together: files updates and database updates. Your point #2 could solve the files updates, assuming that the package is being maintained properly, but for the database stuff, the only way that I know is to do it was manually through the interface (this may have changed with https://plugins.matomo.org/ExtraTools).

As for #3, you can inject the config.ini.php file into your ansible playbooks, no need to manually run through it. :)

My two cents.

**Comment by chriscroome** [comment]
The Matomon Ansible role I have been working on and using is here: https://git.coop/webarch/matomo

I've been using it for upgrading quite a few sites and have recently also got it working for installs and adding users, the next step is adding sites, but that is currently held up by [this issue](https://github.com/digitalist-se/extratools/issues/9).

**Comment by J0WI** [comment]
I tried this in https://github.com/J0WI/docker-piwik/commit/6015481e1b139d8f10fb10bba4234b9d6d1df8d2 for the [Docker container](https://github.com/matomo-org/docker). This would be very useful.

**Comment by colans** [comment]
> for the database stuff, the only way that I know is to do [upgrades] was manually through the interface

This can now be done by the Debian package.  I ended up using a DB dump template after all, but it doesn't matter if it gets stale.  Once the Debian package gets installed (at the end), it'll upgrade the DB schema.  And of course, this will happen whenever OS updates are run.

See [my role](https://gitlab.com/consensus.enterprises/ansible-roles/ansible-role-matomo) for details, soon to be published on [Galaxy](https://galaxy.ansible.com/).

So I didn't need any of these console commands yet, but I'm certainly glad folks have been working on this as I'm sure I'll find it useful later (e.g. for adding users, etc.).  Please keep up the great work!

*Edit: My role is now published at https://galaxy.ansible.com/consensus/matomo.*

**Comment by AbcSxyZ** [comment]
I saw https://github.com/matomo-org/matomo/issues/1586 and this issue. I was also intersting by this kind of feature, and I'm not sure if someone is working on, issue of this kind have years, checking current commit, seem having nothing...

I'm working on since 2 days and it's still in progress, but I'm trying to code within the Installation plugin to add an headless interface to be filled with variable.

At this day, I've been able to extract the installation logic from the plugin controller, and go through all installation step. All setup feature are done in my `InstallManager` : https://github.com/AbcSxyZ/matomo/blob/headless_init/plugins/Installation/InstallManager.php

At this day, I've no way to send my own data, but It seem I've got the API to do it. I've to care about error and some spec, and think about the way to check if the install have to be done headless or not.

I'm thinking it in a way that could be isolated, to avoid breaking feature in Installation plugin, but in a way that installation controller can use it for upgrade, because there is a lot of change. It's possible easily because I had to isolate the installation entirely contained in the Controler (see https://github.com/AbcSxyZ/matomo/blob/headless_init/plugins/Installation/Controller.php).

Any feedback is welcome, and if there is any tips to increase the php style because I'm not using it usually.

**Comment by chriscroome** [comment]
@AbcSxyZ I have implemented [automated installs both via git and the tgz download using Ansible](https://git.coop/webarch/matomo/-/blob/master/tasks/install.yml) and the [Extra Tools matomo:install command](https://github.com/digitalist-se/extratools), see [this PHP file](https://github.com/digitalist-se/extratools/blob/master/Commands/InstallMatomo.php).

**Comment by AbcSxyZ** [comment]
I thought this tool was a Web Interface wrapper, because it wasn't added to the core, but it's using piwik functions to perform the install. Also https://github.com/nebev/piwik-cli-setup/blob/master/install.php have the same kind of setup. I'm wondering why they didn't added it to the matomo source code.

It look intersting to me to add this features in the Installation plugin.

**Comment by AbcSxyZ** [comment]
I'm able to go through the admin installation and redirect to default `index.php` using an array, having all table created. See `welcome` function of the `Installation\Controller` to get the array and my `headlessInstall` in my `InstallManager` for the installation.

**Comment by AbcSxyZ** [comment]
**UPDATE**
I'm getting a mecanisme which work like that :
- Create a `config.ini.php`
- Setup a `MatomoInstall` section, and filled like so: https://github.com/AbcSxyZ/matomo/blob/headless_init/config/config.ini.php.template

This will install the matomo server on first request, and show the login page of matomo. I had to change some core feature, like `isPiwikInstalled` and the `EnvironnementValidator` to plug the InstallManager and perform some test. I'm not aware of the best possible place, I tried to find it, at least for testing.

I'm not sure if all is running properly (fore sure not), but I'm able to create a correct config file, fill database, login into the matomo homepage with specified password. I tried to activate and deactivate a plugin, it works. I've no way to test it with a Matomo normal usage, I've no website using it, and I'm still stuck at the installation !

**Comment by bluikko** [comment]
As is evident from the amount of issues opened around this problem - any modern app these days really must support headless install.

The days of manually browsing to an install page are truly over.

**Comment by chriscroome** [comment]
While the [ExtraTools plugin doesn't work for Matomo 4 yet](https://github.com/digitalist-se/extratools/issues/21), so the console command `matomo:install` can't be used for automatic installs, I have found that automating via the web installer is not so hard, see for example [these Ansible tasks](https://git.coop/webarch/matomo/-/blob/370003b96c1728461167127a03d4a5fdc1d08805/tasks/install_web.yml) which are [based on these tasks](https://github.com/Aversiste/ansible-matomo/blob/master/tasks/main.yml).

However I do agree with @bluikko — it would be great to have the ability to properly interact with Matomo using a CLI.

**Comment by gitressa** [comment]
Just chiming in here, and agreeing that having Matomo CLI installation would be awesome.

You can install Drupal with Drush from the command line with this:

`drush site:install --db-url=mysql://drupal9:drupal9@database/drupal9`

It would be great if similarly you could download Matomo, instantiate whatever web server you like and install it:

```
matomo install \
--db-url=mysql://lamp:lamp@database/lamp \
--option user=admin \
--option pw=admin \
--option email=admin@example.org \
--option sitename='Example Org' \
--option siteurl=example.org
```

**Comment by arosenhagen** [comment]
FYI: we managed to automate the installation in a docker environment without going through the web-wizard by
a) cloning the matomo github repo
b) cloning the ExtraTools-Plugin into the plugin directory
c) update composer dependencies (incl. sub-repositories)
d) build a custom docker image (especially requiring mysql-client / mysqladmin dependencies for ExtraTools)
e) running the parameterized install command from ExtraTools in the container

We also added several other plugins and install those properly by ExtraTools.

**Comment by vikenbauer** [comment]
@arosenhagen , how do you create the database? I've tried to repeat the steps you tell above, but it doesn't allow to activate ExtraTools as there is no database created (and I've to activate ExtraTools for being able to run step e ).


> FYI: we managed to automate the installation in a docker environment without going through the web-wizard by a) cloning the matomo github repo b) cloning the ExtraTools-Plugin into the plugin directory c) update composer dependencies (incl. sub-repositories) d) build a custom docker image (especially requiring mysql-client / mysqladmin dependencies for ExtraTools) e) running the parameterized install command from ExtraTools in the container
> 
> We also added several other plugins and install those properly by ExtraTools.

**Comment by bluikko** [comment]
It does work but it is very, very painful.

The most stupidest thing of all (...I've ever seen?!) is that the database is not fully initialized until first HTTP GET to the website!!

**Comment by vikenbauer** [comment]
Following @arosenhagen message https://github.com/matomo-org/matomo/issues/10257#issuecomment-1018498246 guidance, I've managed to get a full unattended installation, doing some tinkering with ExtraTools (https://plugins.matomo.org/ExtraTools). Those are the steps I've followed:

1. Choose your favourite method for getting a uninitialized Matomo server (in my case, I was using a docker compose configuration like https://github.com/matomo-org/docker/blob/master/.examples/apache/docker-compose.yml). I've not tried installing from zip or git in a PHP server, but I guess it doesn't change so much.
2. Install ExtraTools server dependencies:
`apt update && apt install -y unzip mariadb-client`
(Dependency is actually mysqladmin, but for mariaDB it comes into mariadb-client package).
3. Install `composer`.
4. If Matomo `composer.json` is missing, add it (i.e. from GitHub: https://raw.githubusercontent.com/matomo-org/matomo/4.7.1/composer.json ).
5. Install ExtraTools composer dependencies: https://plugins.matomo.org/ExtraTools#documentation.
6. Install ExtraTools: Download zip, uncompress and move it to plugins directory.
7. Copy a `config.ini.php` in Matomo config dir, containing database connection parameters only. 
8. Copy `plugins/ExtraTools/Commands/InstallMatomo.php` to `plugins/CoreConsole/Commands` and change its namespace to `Piwik\Plugins\CoreConsole\Commands`: Here is where the magic is. ExtraTools is not going to work initially; there is no database created yet so ExtraTools activation fails. Doing this copy, the next step will work.
10. Run `matomo:install` command: Installation parameters can be added as command arguments, but it's easier to create an `install.json` file (like in https://github.com/digitalist-se/extratools/blob/4.x-dev/docs/example.json):
`console matomo:install --install-file=install.json` 
11. Delete temporary `plugins/ExtraTools/Commands/InstallMatomo.php.`
12. Add an initial site (current ExtraTools version doesn't create any site in matomo:install).
13. As I'm building a docker container, cleaning up (mariadb-client, ExtraTools, composer...).

All steps can be easily automated in a script or Dockerfile.

I hope it helps someone there. As I said is mostly tinkering, and I think this feature would have to be available in vanilla Matomo, as unattended containers or virtualized installations are common today.

**Comment by CrimsonScythe** [comment]
Hey @vikenbauer I am trying to get this set up on a serverless system. Therefore, I am extending the Docker image for matomo:alpine and writing the steps you mentioned in that Dockerfile. 
However, I seem to be getting this error when running `matomo:install`:
`[Exception]                                                                                
  Adapter 'mysql' is not valid. Maybe check that your Matomo configuration files in config/  
  *.ini.php are readable by the webserver.  `
 I have given the folder read and write access to everything in config/ and also doublechedk with `ls -l` any ideas on what could be wrong?
 
 Trying to activate the ExtraTools plugin leads to the following error:
 `[PDOException]                                                                    
  SQLSTATE[42S02]: Base table or view not found: 1146 Table 'matomo.matomo_option'  
   doesn't exist    `

**Comment by CrimsonScythe** [comment]
I updated the confg.ini.php to only have db connection params and the plugin now activates however, still getting the same error on `matomo:install`. 
On closer inspection, `matomo:install` seems to be changing the config.ini.php file to include `adapter = "mysql"` under [database]

**Comment by CrimsonScythe** [comment]
On closer inspection, `matomo:install` seems to be changing the config.ini.php file to include `adapter = "mysql"` under [database]

**Comment by JMLX42** [comment]
> All steps can be easily automated in a script or Dockerfile.

@vikenbauer do you mind providing a bash script to do all of that please?

**Comment by its-Saab** [comment]
I'm running Matomo 4.12.3 in Swarm and all tables are created in DB. I've added environment variables for the DB related fields to the stack but whenever the container restarts I have to go through the wizard (but with prefilled credentials from the environment variables). you mentioned having config.ini.php beforehand will allow us to skip the installation wizard, could you please provide an example of how to set that up in a swarm env?

would be a good example to add to your documentation unless it already exists and I failed to find it

**Comment by michalkleiner** [comment]
Hi @its-Saab,

thank you for the question. Unfortunately we are not in a position to provide custom setup support, however we would happily review a community raised PR.

In multi-server environments, it is important that all containers have access to the shared filesystem where the config file is written to and that it is persisted across container restarts. Once you configure your Matomo instance, if you locate which container (and where) created `config.ini.php`, you might be able to either copy the file to other containers or use it for when containers restart, though as I mentioned at the beginning, the best solution is to make sure the files created by Matomo are persisted and available to new containers. With that, all new containers should pick it up automatically. If that's not possible, you may make `config/confg.ini.php` part of your container build.

**Comment by its-Saab** [comment]
@michalkleiner thanks for your response!

I ended up using "volumes" where the initial content of that volume is created by a Matomo container which seems to have solved my issue. 

The thing is, at least when I was looking into it, I found some examples mentioned in the docs but not pointing this requirement out as you just did

**Comment by achangjvi22** [comment]
Just a note on our side, if you're all interested in running in kubernetes, I ended up creating a helm chart mapping config.ini.php, /tmp, matomo.js and the GeoIP mmdb in /misc to a NFS persistent volume. and I'm able to run matomo as a replicaset. 

it took me a while to get to this point, but I had been migrating my multiple Matomo VM's to Kubernetes, so didn't have to do the setup part while in a container.... However, this thread has given me insight in to programmatically setting up a new instance. so I'm gonna give that a try..

thanks all..!

**Comment by jsabater** [comment]
Does anyone know whether the console:install command provided by ExtraTools is still missing from Matomo 5.x? I am about to use [Víctor's work](https://github.com/matomo-org/matomo/issues/10257#issuecomment-1039352193) on my Ansible playbook but it would be lovely not having to run through so many steps but, instead, have it native, in the core.

@vikenbauer , are you still using the same procedure these days?

**Comment by chriscroome** [comment]
I don't know the answer to that @jsabater -- I switched to using the [web installer via Ansible](https://git.coop/webarch/matomo/-/blob/master/tasks/install_web.yml) and it's been working fine for some years!

**Comment by jsabater** [comment]
I just finished both set of tasks on my Ansible playbook today, @chriscroome , and I will be testing installations from scratch tomorrow (today I was only testing upgrades). I'll get back to you as soon as possible.

**Comment by palminha** [comment]
any news on this?

this is important for setting up matomo in more automated way, like containers.

currently if use container we always need to a persistent volume, because independently of setting the DATABASE credentials and URL and ENV variables, matomo ignores and always show the setup screen. This avoids deploying matomo in a more autonomous way (like using AWS AppFlow)

**Comment by jsabater** [comment]
> I don't know the answer to that [@jsabater](https://github.com/jsabater) -- I switched to using the [web installer via Ansible](https://git.coop/webarch/matomo/-/blob/master/tasks/install_web.yml) and it's been working fine for some years!

So, in the end the tests didn't work and I am, currently, using the same system as you: the [web installer via Ansible](https://git.coop/webarch/matomo/-/blob/master/tasks/install_web.yml).

**Comment by mikkeschiren** [comment]
> any news on this?
> 
> this is important for setting up matomo in more automated way, like containers.
> 
> currently if use container we always need to a persistent volume, because independently of setting the DATABASE credentials and URL and ENV variables, matomo ignores and always show the setup screen. This avoids deploying matomo in a more autonomous way (like using AWS AppFlow)

@palminha  Have you tried Extra Tools? We are using it for containers in Kubernetes.

**Comment by palminha** [comment]
> > any news on this?
> > this is important for setting up matomo in more automated way, like containers.
> > currently if use container we always need to a persistent volume, because independently of setting the DATABASE credentials and URL and ENV variables, matomo ignores and always show the setup screen. This avoids deploying matomo in a more autonomous way (like using AWS AppFlow)
> 
> [@palminha](https://github.com/palminha) Have you tried Extra Tools? We are using it for containers in Kubernetes.

since i'm using AWS, i'm planning to use AWS BeanStalk

**Comment by mikkeschiren** [comment]
@palminha - I mean: https://plugins.matomo.org/ExtraTools

**Comment by mikkeschiren** [comment]
So with ExtraTools you can install Matomo with the console.

**Comment by bluikko** [comment]
> So with ExtraTools you can install Matomo with the console.

Does this still work in the latest versions? Last time I tested it I think it only worked on some older versions but I may be mistaken, it's been a while.  
I believe there has been talks about the better way being to just drive the stock installation web pages with sending manual HTTP requests.

**Comment by mbrodala** [comment]
It is working all the time, we have automated tests which ensure this.

It basically replicates what the `plugins/Installation/Controller.php` does but in a CLI environment. So there is no need to trigger web requests for the installation.
